File: android/app/src/main/java/com/touliao/app/core/push/PushRegistrationQueue.kt
package com.touliao.app.core.push

import com.touliao.app.core.storage.TokenStore
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

internal class PushRegistrationQueue(private val credentials: TokenStore) {
    private val mutex = Mutex()

    // A switch must drain an in-flight registration before removing its ownership.
    suspend fun run(owner: TokenStore.Snapshot, operation: suspend () -> Unit): Boolean = mutex.withLock {
        if (owner.token.isNullOrBlank() || !credentials.isCurrent(owner)) return@withLock false
        operation()
        true
    }
}

File: ios/Touliao/Core/Push/PushRegistrationQueue.swift
import Foundation

actor PushRegistrationQueue {
    private let credentials: KeychainStore
    private var tail: Task<Void, Never>?
    init(credentials: KeychainStore = .shared) { self.credentials = credentials }

    // Actors alone are reentrant across await. Chain tasks to drain an in-flight
    // registration before the switch deletes the old session's destinations.
    func run(owner: KeychainStore.Snapshot, operation: @escaping () async -> Void) async {
        let previous = tail
        let task = Task {
            await previous?.value
            guard owner.token?.isEmpty == false, credentials.isCurrent(owner) else { return }
            await operation()
        }
        tail = task
        await task.value
    }
}
